### ISSUE TITLE
"Your browser doesn't support WebGL" error in PBI Desktop

### ISSUE THREAD

**Issue by vitaly-livshits** [mapbox/mapboxgl-powerbi#71]
We are getting this error in PBI Desktop with the latest (180219) release of the MapBox visualization. Running latest PBI Desktop version (2.55.5010.641). The machine is Windows 10 and has IE, Edge and latest Chrome installed.

How to troubleshoot this?

![image](https://user-images.githubusercontent.com/1765732/36738791-6701331c-1bac-11e8-9cca-f2b0e3e94c60.png)

**Comment by ryanbaumann** [comment]
@vitaly-livshits thanks for the error report.  

Does loading this debug page on the same Windows 10 machine you're seeing this error in PowerBI Desktop on https://www.mapbox.com/mapbox-gl-js/example/check-for-support/ in any of your browser windows throw an alert?

**Comment by otravers** [comment]
Not sure Power BI even cares so this just an idle theory: try changing the default browser to Chrome.

**Comment by vitaly-livshits** [comment]
Yes, this is definitely environment specific. 

The machine with the issue is equipped with a DisplayLink-based docking station and is running the latest DisplayLink drivers. Do you have any DisplayLink-based docks available at Mapbox? 

I only have intermittent access to the environment. I will run the suggested tests the next time I have access and report the results here.

**Comment by vitaly-livshits** [comment]
Just tried - https://www.mapbox.com/mapbox-gl-js/example/check-for-support/
page loads fine in all 4 browsers installed on the machine.
http://webglreport.com/ confirms that WebGL 1 is available in all 4
browsers. The PC has a GPU, it's Intel HD Graphics 520.

There must be something particular about rendering from within PBI Desktop.

Any ideas?

**Comment by ryanbaumann** [comment]
This may be specific to how Power BI Desktop embedded browser works on this machine @vitaly-livshits - let's see what the reply is on https://github.com/Microsoft/PowerBI-visuals/issues/310

**Comment by Molotch** [comment]
I have the same issue in a Azure Windows 10 Pro Iaas VM.

http://webglreport.com/ confirms Edge supports WebGL 1.0 and the map renders just fine at https://www.mapbox.com/mapbox-gl-js/example/check-for-support/

Power BI desktop shows the error.

**Comment by ryanbaumann** [comment]
Thanks for the error report @molotch!  I'll continue to work on how to debug this better with the Power BI team.

**Comment by zapcity** [comment]
I have exactly the same issue.

I am on a Dell Latitude 7480 with Intel HD Graphics 520

Tested latest versions of chrome, edge, firefox and internet explorer. Using March 2018 version of PowerBI.

Results:

- https://www.mapbox.com/mapbox-gl-js/example/check-for-support/ checks out ok for all 4 browsers

- http://webglreport.com/ confirms WebGL 1 is available in all 4 browsers. WebGL 2 has issues on edge and internet explorer, all the rest are fine.

Looking forward to a fix.

**Comment by ryanbaumann** [comment]
Thanks for the additional bugs these are very helpful.  I can cut a test version to try locally and confirm it solves your issue.  Ill drop an update tomorrow.

**Comment by ryanbaumann** [comment]
@zapcity @woodwind3442 @Molotch @vitaly-livshits 

I created a test build that removes the check we perform to see if the browser self-reports if WebGL is supported. Please try the [test pbiviz file](https://github.com/mapbox/mapboxgl-powerbi/raw/no-webgl-validate/dist/mapboxGLMap.pbiviz) and report back on this ticket if this fixes your issue.

[Reference Pull Request.](https://github.com/mapbox/mapboxgl-powerbi/pull/83)
 My hunch here is that the browser is reporting that webGL is not supported, but the browser has an integrated software rendering fallback.

**Comment by vitaly-livshits** [comment]
Just tried the test build. The issue is still there.

**Comment by zapcity** [comment]
Same, issue still persists. I deleted the old custom vis and installed the new one via the link given. Also used my access token. I'll give this a try on my desktop when I get home tonight. It just doesn't work on the work laptop with Intel HD Graphics 520.

**Comment by ryanbaumann** [comment]
Do you get the exact same error (`the browser does not support WebGL`) as before @vitaly-livshits @zapcity?  Or does the map just not render (blank div) with the new version?

**Comment by vitaly-livshits** [comment]
I am getting the exact same error as before.

**Comment by ryanbaumann** [comment]
@vitaly-livshits how does this work?  https://github.com/mapbox/mapboxgl-powerbi/raw/no-webgl-validate/dist/mapboxGLMap.pbiviz

**Comment by vitaly-livshits** [comment]
This one works, thanks!

**Comment by ryanbaumann** [comment]
Woho @vitaly-livshits!  

@zapcity @woodwind3442 @Molotch can you confirm this solves your issue as well?

**Comment by zapcity** [comment]
Thanks we are making progress! I no longer get the "Your browser doesn't support WebGL" error in my mapbox visualisation. However, I get a progress circle and it keeps generating an error dialog box saying:

> Something went wrong: Cannot read property 'resize' of undefined

I have the default map visualization working ok with my lat/long data, but just can't get the mapbox visualisation to work.

**Comment by ryanbaumann** [comment]
@capacity it looks like the Intel HD 520 card has a history of WebGL errors with Chrome.  Chrome is the browser Power BI Desktop users. https://communities.intel.com/thread/117565.  Are you using thr latest graphics card drivers?

Can you open the sample Mapbox Visual for Power BI workbooks from the readme in this repo in Chrome and Firefox?  Do they show errors?

**Comment by jamesdales** [comment]
I also get this issue. With the current App Source version I get "your browser doesn't support WebGL". With the latest build linked above, I get a never ending loading circle once I put in the mapbox key.

This all happens ONLY when I'm plugged into any DisplayLink USB hub, I've tried three different makes and they all suffer the same issue. I'm using the latest drivers.  It works fine without the DisplayLink device connected. Once PBI has loaded and I add the mapbox visual, then I can plug in the display link device and it's fine. 

This only happens in Power BI desktop, it doesn't happen in the browser - Edge or Chrome.

**Comment by zapcity** [comment]
@ryanbaumann Thanks for the update. Ok, I've done a fair bit of testing. I updated all drivers for my laptop from the Dell site. I can confirm that the sample mapbox visuals on the github repo site work in both firefox and chrome. However when I download the repo and run the pbix files in the \examples directory e.g. 'NYC Cycling Incident Dashboard.pbix' I get the following:

**With my laptop connected to my DisplayLink USB hub**
1. When I open 'NYC Cycling Incident Dashboard.pbix' it gives me "Your browser doesn't support WebGL". This workbook is obviously using the old pbiviz, as it was last updated a month ago.
2. I install the new pbiviz you linked to. I don't get the "Your browser doesn't support WebGL", but instead I get a spinning waiting circle.

**With my laptop disconnected from my DisplayLink USB hub**
1. When I close/reopen the 'NYC Cycling Incident Dashboard.pbix' it works straight out the box i.e. I don't need to install the updated pbiviz.

So I can confirm exactly what @jamesdales said. I have tested the DisplayLink USB hub with ethernet and wifi connectivity, and it makes no difference. Hope that helps in the troubleshooting.

**Comment by zapcity** [comment]
Update to the comments I made above today. If I want to display the pbix on my external monitors, I have to: 

1. Firstly be disconnected from the DisplayLink USB hub. 
2. Open the pbix. 
3. Then connect to the DisplayLink USB hub

**Comment by zapcity** [comment]
Ok, I got it working on the default 'NYC Cycling Incident Dashboard.pbix' and my custom pbix. The issue was that I had to update the DisplayLink drivers from http://www.displaylink.com/downloads/windows

The DisplayLink drivers are not a part of the Dell suite of updates when downloading from http://www.dell.com/support/home/au/en/aubsd1

Thanks again @ryanbaumann for all your help.

**Comment by zapcity** [comment]
I think I got it working. I spent A LOT of time today testing different things. I installed all the latest drivers from Dell. Also updated the DisplayLink drivers from displaylink.com. This is what it comes down to:

- If I set my laptop as 'Make this my main display' in windows Settings > Display. The pbix with mapbox pbiviz opens up without error in my laptop screen by default. I can then drag it to my external screen without issue. Note, in this case, the old pbiviz seems to be working as I tested it on 'NYC Cycling Incident Dashboard.pbix' found in the examples folder of the repo. This file was last updated 1 month ago.
- If I set up one of my external monitors as 'Make this my main display', the pbix opens in the external display by default, and the pbixviz has the usual errors as documented in my previous posts. This setting will emulate the situation @jamesdales seems to be having. Note, the _'Something went wrong: Cannot read property 'resize' of undefined'_ error I stated in my previous post occurs in this situation, but I finally realised it was happening when resizing the pbiviz window with the mouse.

So I guess I just have to keep my laptop as my main display - something I prefer not to do. But hey, it works now. So I'm happy. Thanks @ryanbaumann for all your help.

**Comment by ryanbaumann** [comment]
Thanks for the detailed instructions on how to debug and fix, @zapcity!  Glad it's working now. Sorry to hear it took quite a bit of troubleshooting - hardware driver issues are always fun to debug.  

TL;DR Mapbox GL JS requires `Web GL` support.  It sounds like there is a hardware driver specific issue with using an external monitor supporting WebGL in PowerBI Desktop.  It's worth a shot to open this up in the Power BI Visuals repository if the Power BI Online version of the viz works with the exact same setup.

**Comment by zapcity** [comment]
Thanks @ryanbaumann. Done: https://github.com/Microsoft/PowerBI-visuals/issues/320

**Comment by ryanbaumann** [comment]
I removed the WebGL error message check in https://github.com/mapbox/mapboxgl-powerbi/pull/76

**Comment by Pimsen** [comment]
Hi Ryan,
I have the same issues on different devices (mobiles, tablets as well as laptops).

I tried your test build. In power bi the visual is not show the error message anymore but its loading all the time (the spinning circle). If I publish the report, the visual is working fine! Are there any restrictions in your test build?

Kind regards from Zurich
Steve

**Comment by ryanbaumann** [comment]
Hi @pimsen - what browser are you using locally when you only see the spinning circle?

**Comment by Pimsen** [comment]
Hi @ryanbaumann - Thank you very much :)

I have different browsers. Mainly I use Chrome. Its my standard Browser.

The map is loading (spinning circle) only in power bi desktop. In the browser its working.
When I configure the map it's necessary to publish to check the changes. It takes time to getting work done. It would be nice if we can check the map in the desktop application not after publishing.

**Comment by cdtavijit** [comment]
I still can't get this to work. I have enabled webgl in my chrome and restarted. However, it still does not work for my case.

**Comment by ryanbaumann** [comment]
cc/ @samgehret

**Comment by brentlightsey** [comment]
I seem to still get this message any time I'm using a virtualized PC. I've seen it happen in VMWare VPCs as well as one running on Azure. Thought that may be useful to know.
